File: Microsoft/147.insertion-sort-list.java
/**
 * Algorithm:
 * Build a sorted linked list by inserting each original node into its correct
 * position after a dummy head. This is direct insertion sort on linked nodes.
 *
 * Complexity:
 * Time O(n^2), space O(1).
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode dummy = new ListNode(0);
        ListNode cur = head;
        while (cur != null) {
            ListNode next = cur.next;
            ListNode prev = dummy;
            while (prev.next != null && prev.next.val < cur.val) {
                prev = prev.next;
            }
            cur.next = prev.next;
            prev.next = cur;
            cur = next;
        }
        return dummy.next;
    }
}

File: Bloomberg/147.insertion-sort-list.go
package leetcode

// InsertionSortList147 inserts each original node into a new sorted list headed
// by a dummy node. The list is singly linked, so each insertion scans from dummy
// to find the first node with value >= current.
//
// Time: O(n^2). Space: O(1).
func InsertionSortList147(head *ListNode) *ListNode {
	dummy := &ListNode{}
	for cur := head; cur != nil; {
		nxt := cur.Next
		prev := dummy
		for prev.Next != nil && prev.Next.Val < cur.Val {
			prev = prev.Next
		}
		cur.Next = prev.Next
		prev.Next = cur
		cur = nxt
	}
	return dummy.Next
}
